---
title: "DIMA report"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
date: "DATE"
author: 'Yinxiu Zhan'

params:
  patientid: NULL
  annotation_filters: NULL
  metrics_file: NULL
  variants_somatic_file: NULL 
  variants_germline_file: NULL 
  clonal_tmb_file: NULL 
  variant_signatures_file: NULL 
  dragen_tmb_file: NULL 
  dragen_msi_file: NULL
  rna_biomarker: NULL
  cnv_file: NULL

---

```{r setup}
knitr::opts_chunk$set(echo = FALSE, results = "asis", message = T, warning = FALSE)
```

```{r function}
add_biomarker = function(result, input){
    if(!is.null(input)){
      colnames(input) = c("biomarker", "value")
      return(rbind(result,input))
    }else{
      return(result)
    }
}

add_variants = function(result, input, type){
  select = c(
      "Hugo_Symbol",
      "cDNA_Change",
      "Protein_Change",
      "tumor_f",
      "type"
    ) 
  if(!is.null(input)){
      input$type = type
      select = select[select %in% colnames(input)]
      input = input[,select]
      return(rbind(result,input[,select]))
  }else{
    return(result)
  }
}

isvalid = function(var){
  return(!is.null(var) && var!="null")
}
```

```{r library}
library(DT)
library(knitr)
library(maftools)
library(rjson)
library(stringi)
library(stringr)
```


```{r gather data, echo=FALSE, results=FALSE, include=FALSE}
variants_somatic = NULL
variants_germline = NULL
clonal_tmb = NULL
variant_signatures = NULL
dragen_msi = NULL
dragen_tmb = NULL
cnv = NULL

if(isvalid(params$variants_somatic_file)) {
  variants_somatic = read.maf(params$variants_somatic_file)
  variants_somatic = variants_somatic@data
  variants_somatic = variants_somatic[variants_somatic$filter == "PASS", ]
  variants_somatic = as.data.frame(variants_somatic)
  if(nrow(variants_somatic) == 0){variants_somatic = NULL}
}

if(isvalid(params$variants_germline_file)) {
  variants_germline = read.maf(params$variants_germline_file)
  variants_germline = variants_germline@data
  variants_germline = variants_germline[variants_germline$filter == "PASS", ]
  variants_germline = as.data.frame(variants_germline)
  if(nrow(variants_germline) == 0){variants_germline = NULL}
}

if(isvalid(params$clonal_tmb_file)) {clonal_tmb = read.table(params$clonal_tmb_file, sep=":")}
if(isvalid(params$variant_signatures_file)) {variant_signatures = read.table(params$variant_signatures_file, sep=":")}
if(isvalid(params$dragen_tmb_file)) {dragen_tmb = read.csv(params$dragen_tmb_file, header=FALSE)}
if(isvalid(params$dragen_msi_file)) {dragen_msi = fromJSON(file=params$dragen_msi_file)}
if(isvalid(params$cnv_file)) {cnv = read.delim(params$cnv_file)}
```




```{r gather metadata, echo=FALSE, results=FALSE, include=FALSE}
patientid_somatic=NULL
patientid_germline=NULL

if(!is.null(variants_germline)){
  if(ncol(variants_germline)>1){
    variants_germline$tumor_f = round(variants_germline$tumor_f, 3)
    patientid_germline = unique(variants_germline$Matched_Norm_Sample_Barcode)
  }
}

if(!is.null(variants_somatic)){
  if(ncol(variants_somatic)>1){
    variants_somatic$tumor_f = round(variants_somatic$tumor_f, 3)
    patientid_somatic = unique(variants_somatic$Tumor_Sample_Barcode)
  }
}
```


# Whole exome sequencing

The whole exome sequencing data have been processed using Illumina DRAGEN Bio-IT Platform (v.4.2.4). 
The variants have been annotated and prioritized using [variantalker](https://github.com/zhanyinx/variantalker) v1.4

The analysed samples are:

- somatic sample: `r patientid_somatic`

- germline sample: `r patientid_germline`

## Small Variants

The list of variants passing the project filtering parameters (see Appendix) is: 

```{r}
variants = NULL

variants = add_variants(variants, variants_somatic, "somatic")
variants = add_variants(variants, variants_germline, "germline")

if(!is.null(variants) && ncol(variants)>1){
  DT::datatable(variants, filter="top", rownames=FALSE)
}
```

## Biomarkers

We used different tools to extract biomarkers from whole exome sequencing and RNA-seq data. In particular, 
Illumina DRAGEN Bio-IT Platform allows the extract of MSI and TMB from WES data. We then complemented the list of biomarkers
using the biomarker pipeline from variantalker (see [here](https://github.com/zhanyinx/variantalker/tree/main/docs/biomarkers)).


Recommendations: MSI instability if percentage >= 20 for solid tumor; 

```{r biomarkers}
biomarkers = NULL
biomarkers = add_biomarker(biomarkers, clonal_tmb)
biomarkers = add_biomarker(biomarkers, variant_signatures)

if(length(dragen_msi)>1){
  biomarkers = add_biomarker(biomarkers, data.frame(v1="dragen_msi", v2=dragen_msi$PercentageUnstableSites))
}

if(length(dragen_tmb)>1){
  dragen_tmb = dragen_tmb[c(nrow(dragen_tmb)-1, nrow(dragen_tmb)),c(ncol(dragen_tmb)-1, ncol(dragen_tmb))]
  dragen_tmb[,1] = paste0("dragen_",dragen_tmb[,1])
  biomarkers = add_biomarker(biomarkers, dragen_tmb)
}

biomarkers$value = round(as.numeric(biomarkers$value), 3)

knitr::kable(
  biomarkers,
  booktabs = TRUE, valign = 't',
  row.names=FALSE
)
```

## Copy number variations

We prioritized the CNV using [2019 ACMG classification guidelines](https://www.sciencedirect.com/science/article/pii/S1098360021013009?via%3Dihub)

Below all the of genes within CNVs defined as pathogenic or likely pathogenic according to the guidelines. 

Probably it's better to have a list of genes we wanna check.
```{r cnv}
keep = c("Likely pathogenic", "Pathogenic")

if(length(cnv)>1){
  filtered = cnv[cnv$Classification %in% keep, ]
  
  if(nrow(filtered) > 0){
    dup = filtered[filtered$Type=="DUP",]
    del = filtered[filtered$Type=="DEL",]
    
    loss = str_squish(unlist(strsplit(del$All.protein.coding.genes, ",")))
    gain = str_squish(unlist(strsplit(dup$All.protein.coding.genes, ",")))
    
    if(length(loss)>0){
      loss = data.frame(gene = loss, type="loss")
    }else{
      loss = NULL
    }

    if(length(gain)>0){
      gain = data.frame(gene=gain, type="gain")
    }else{
      gain = NULL
    }
    all = rbind(loss,gain)
    
    DT::datatable(all, filter="top", rownames=FALSE)
  }
}
```


# Appendix {#appendix}

## Coverage

For each sample, Dragen provides statistics on coverage. Below the coverage for all samples included in the report

```{r coverage}
if(!is.null(params$metrics_file)){
  cleaned_string <- gsub("\\[|\\]", "", params$metrics_file)
  single_paths <- unlist(strsplit(cleaned_string, ", "))

  for(file in single_paths){
    name = unlist(strsplit(basename(file), "\\."))[1] 
    metrics_data = read.csv(file, header=FALSE)[,c(3,4)]
    colnames(metrics_data) = c("metrics", "value")
    metrics_data$value = as.character(metrics_data$value)
    metrics_data = metrics_data[-c(1,2,6,7,8,9,10,11,12,13,14,15,16),]
    print(kable(metrics_data, caption=name))
  }
}
```



## meta data 

```{r database, echo=FALSE, results=FALSE, include=FALSE}
somatic_annovar=NULL
somatic_funcotator=NULL

germline_annovar=NULL
germline_funcotator=NULL

if(isvalid(params$variants_somatic_file)) {
  command=paste0('grep "  Funcotator" ', params$variants_somatic_file)
  somatic_funcotator <- system(command, intern = TRUE)
  somatic_funcotator = gsub("#", "", somatic_funcotator)

  command=paste0('grep "Annovar" ', params$variants_somatic_file)
  somatic_annovar <- system(command, intern = TRUE)
  somatic_annovar = gsub("#", "", somatic_annovar)
}

if(isvalid(params$variants_germline_file)) {
  command=paste0('grep "  Funcotator" ', params$variants_germline_file)
  germline_funcotator <- system(command, intern = TRUE)
  germline_funcotator = gsub("#", "", germline_funcotator)

  command=paste0('grep "Annovar" ', params$variants_germline_file)
  germline_annovar <- system(command, intern = TRUE)
  germline_annovar = gsub("#", "", germline_annovar)
}
```


The list of databases used for the annotation is the following:

Somatic:

- `r somatic_funcotator`

- `r somatic_annovar`

Germline:

- `r germline_funcotator`

- `r germline_annovar`

The variants shown here pass the following list of filters: `r params$annotation_filters`. 

For more info on filters, please have a look [here](https://github.com/zhanyinx/variantalker/tree/dev#output)
